manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java

```diff
@@ -0,0 +1,639 @@
+        if(list.isEmpty()) return Optional.empty();
+
+        //What we do now is, we will try to figure out the latest datapoint where the predicate fails,
+        //before the newValue.
+        //This means that, the state change took place between the datapoint we just found and its next one.
+
+        //Find the first datapoint that passes the negated predicate
+
+        AssetDatapoint StateChangeAssetDatapoint = null;
+
+        try {
+            for (int i = 0; i < list.size()-1; i++) {
+                // Not using Object.equals, but Datapoint.equals
+
+                AssetDatapoint currentDp = list.get(i);
+                AssetDatapoint theVeryPreviousDp = list.get(i+1);
+
+    //            So, if the currentDp passes the predicate,
+                boolean currentDpTest = pred.value.asPredicate(timerService::getCurrentTimeMillis).test(currentDp.getValue());
+    //            and if the very previous one (NEXT one in the array and PREVIOUS in the time dimension).
+    //            FAILS the predicate,
+                boolean previousDpTest = pred.value.asPredicate(timerService::getCurrentTimeMillis).test(theVeryPreviousDp.getValue());
+    //            A state change happened where the state we are looking for was turned on.
+    //            We want the currentDp.
+
+
+                if(currentDpTest && !previousDpTest){
+                    StateChangeAssetDatapoint = currentDp;
+                    break;
+                }
+            }
+
+            if (StateChangeAssetDatapoint != null){
+                if (!pred.value.asPredicate(timerService::getCurrentTimeMillis).test(StateChangeAssetDatapoint.getValue())){
+                    throw new Exception("Found state change datapoint failed predicate");
+                }
+            }else{
+                throw new Exception("Couldn't find asset state change value");
+            }
+
+        }catch (Exception e){
+            getLogger().warning(e.getMessage());
+            return Optional.empty();
+        }
+
+        //BUG: The Timestamp of an asset datapoint is NOT attribute.getTimestamp()!
+        //This makes it much harder to relate datapoints to relate a duration to the time-distance between two datapoints.
+        //To make this easier, we are going to use some time-padding of 1 second from the gap. It's going to be sufficient to cover some edge-cases.
+
+        //The AssetStateDuration will have a startTime of the found state-change variable, and the endDate will be the previousValue's timestamp, that being the end of the state duration
+        Attribute<?> tripAttr = new Attribute<>("LastTripStartedAndEndedAt", ValueType.ASSET_STATE_DURATION, new AssetStateDuration(
+                new Timestamp(StateChangeAssetDatapoint.getTimestamp()),
+                new Timestamp(previousValue.getTimestamp().get())
+        ));
+
+        tripAttr.setTimestamp(StateChangeAssetDatapoint.getTimestamp());
```

Comment:
... I guess this works too. (see previous comment)

Reply:
Still does not solve the issue, it's not a great way to solve this because I also get some "false" trips